futures-util/src/stream/futures_unordered/mod.rs

```diff
@@ -576,19 +552,26 @@ impl<Fut> Debug for FuturesUnordered<Fut> {
     }
 }
 
-impl<Fut> Drop for FuturesUnordered<Fut> {
-    fn drop(&mut self) {
-        // When a `FuturesUnordered` is dropped we want to drop all futures
-        // associated with it. At the same time though there may be tons of
-        // wakers flying around which contain `Task<Fut>` references
-        // inside them. We'll let those naturally get deallocated.
+impl<Fut> FuturesUnordered<Fut> {
+    /// Clear all the futures associated with this task.
+    pub fn clear(&mut self) {
         unsafe {
             while !self.head_all.get_mut().is_null() {
                 let head = *self.head_all.get_mut();
                 let task = self.unlink(head);
                 self.release_task(task);
             }
         }
```

Comment:
In the `drop`, we can skip resetting some fields because we know that the user will not access the FuturesUnordered again. However, in the `clear`, we will need to reset other fields as well (at least `self.is_terminated.store(false, Relaxed)` is needed).

Reply:
should the `ready_to_run_queue` be reset as well?

Reply:
and, can `self.is_terminated.store(false, Relaxed)` also be called in the `drop` impl, or is that extra overhead?

Reply:
Yes

Reply:
Do you mean `is_terminated.store(true)`, or `false`?

Reply:
I meant `false`. That is the same behavior as `FuturesUnordered::new`:

https://github.com/rust-lang/futures-rs/blob/569106699d179ef4a3fbdf1417d7b772da095d28/futures-util/src/stream/futures_unordered/mod.rs#L133-L137

(I'm hoping that `FuturesUnordered::clear` will work as an efficient version of `mem::replace(futures_unordered, FuturesUnordered::new())`.)

Reply:
Sounds good, updated.